<html>
<head>
<title>Title</title>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
</head>
<body>
Braberg, H.<i>, et al.</i> SALIGN: a web server for alignment of multiple protein sequences and structures. <i>Structural Bioinformatics</i> 2012;28(15):2072-2073.<p>
Carugo, O. and Pongor, S. Protein Fold Similarity Estimated by a Probabilistic Approach Based on C-alpha-C-alpha Distance Comparison. <i>J. Mol. Biol.</i> 2002;315:887-898.<p>
Cliff, M.J.<i>, et al.</i> Conformational Diversity in the TPR Domain-Mediated Interaction of Protein Phosphatase 5 with Hsp90. <i>Structure</i> 2006;14:415-426.<p>
Das, A.K., Cohen, P.T.W. and Barford, D. The structure of the tetratricopeptide repeats of protein phosphatase 5: implications for TPR-mediated protein-protein interactions. <i>The EMBO Journal</i> 1998;17(5):1192-1199.<p>
Everitt, B.S., Landau, S. and Leese, M. Cluster Analysis. Great Britain: Arnold; 2001.<p>
Groves, M.R. and Barford, D. Topological Characteristics of Helical Repeat Proteins. <i>Current Opinion in Structural Biology</i> 1999;9:383-389.<p>
Hammani, K.<i>, et al.</i> Helical repeats modular proteins are major players for organelle gene expression. <i>Biochimie</i> 2014;100:141-150.<p>
Hinds Jr, T.D. and Sanchez, E.R. Protein Phosphatase 5. <i>Int J Biochem Cell Biol.</i> 2008;40(11):2358-2362.<p>
Holm, L. and Sander, C. Protein Structure Comparison by Alignment of Distance Matrices. <i>Journal of Molecular Biology</i> 1993;233:123-138.<p>
Kurochkina, N. Specific sequence combinations at parallel and antiparallel helix-helix interfaces. <i>Journal of Theoretical Biology</i> 2008;225:188-198.<p>
Kurochkina, N.A. and Iadarola, M.J. Helical assemblies: Structure determinants. <i>Journal of Theoretical Biology</i> 2015;369:80-84.<p>
Loeliger, J. Version Control with Git. Sebastopol, CA, USA: O'Reilly; 2009.<p>
Madhusudhan, M.S.<i>, et al.</i> Variable gap penalty for protein sequence-structure alignment. <i>Protein Engineering, Design & Selection</i> 2006;19(3):129-133.<p>
Madhusudhan, M.S.<i>, et al.</i> Alignment of multiple protein structures based on sequence and structure features. <i>Protein Engineering, Design & Selection</i> 2009;22(9):569-574.<p>
Main, E.R.G.<i>, et al.</i> Design of Stable  Alpha-Helical Arrays from an Idealized TPR Motif. <i>Structure</i> 2003;11:497-508.<p>
McMahon, E. Identification and analysis of TPR domains through structure and sequence relationships (Unpublished Report). In.: Institute of Structural and Molecular Biology, Dept. of Biological Sciences, Birkbeck, University of London.<p>
Perry, A.J.<i>, et al.</i> Convergent Evolution of Receptors for Protein Import into Mitochondria. <i>Current Biology</i> 2006;16:221-229.<p>
Petsko, G.A. and Ringe, D. Protein Structure and Function. London, UK: New Science Press Ltd; 2004.<p>
Prasad, B.D., Goel, S. and Krishna, P. <i>In Silico</i> Identification of the Carboxylate Clamp Type Tetratricopeptide Repeat Proteins in <i>Arabidopsis</i> and Rice as Putative Co-Chaperones of Hsp90/Hsp70. <i>PLoS ONE</i> 2010;5(9):e12761.<p>
Sikorski, R.S.<i>, et al.</i> A Repeating Amino Acid Motif in CDC23 Defines a Family of Proteins and a New Relationship among Genes Required for Mitosis and RNA Synthesis. <i>Cell</i> 1990;60:307-317.<p>
Taylor, D. Sams Teach Yourself Unix in 24 Hours. USA: Sams Publishing; 2001.<p>
Teetor, P. R Cookbook. Sebastopol, CA, USA: O'Reilly; 2011.<p>
Ye, Y. and Godzik, A. Flexible structure alignment by chaining aligned fragment pairs allowing twists. <i>Bioinformatics</i> 2003;19(Suppl. 2):ii246-ii255.<p>
Ye, Y. and Godzik, A. Database searching by flexible protein structure alignment. <i>Protein Science</i> 2004;13:1841-1850.<p>
Ye, Y. and Godzik, A. FATCAT: a web server for flexible structure comparison and structure similarity searching. <i>Nucleic Acids Research</i> 2004;32:W582-W585.<p>
Yuzawa, S.<i>, et al.</i> Structural basis for interaction between the conserved cell polarity proteins Inscuteable and Leu-Gly-Asn repeat-enriched protein (LGN). <i>PNAS</i> 2011;108(48):19210-19215.<p>
</body>
</html>
